gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/ImmutableFSJobCatalog.java

```diff
@@ -0,0 +1,129 @@
+package gobblin.runtime.job_catalog;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.List;
+import java.util.Properties;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.typesafe.config.Config;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.runtime.api.JobCatalog;
+import gobblin.runtime.api.JobCatalogListener;
+import gobblin.runtime.api.JobSpec;
+import gobblin.runtime.api.JobSpecNotFoundException;
+import gobblin.runtime.util.FSJobCatalogHelper;
+import gobblin.util.ConfigUtils;
+import gobblin.util.PathUtils;
+import gobblin.util.PullFileLoader;
+
+
+public class ImmutableFSJobCatalog implements JobCatalog {
+
+  protected final FileSystem fs;
+  protected final Config sysConfig;
+
+  protected final JobCatalogListenersList listeners;
+  protected static final Logger LOGGER = LoggerFactory.getLogger(ImmutableFSJobCatalog.class);
+
+  protected final PullFileLoader loader;
+
+  /* The root configuration directory path.*/
+  protected final Path jobConfDirPath;
+  private final FSJobCatalogHelper.JobSpecConverter converter;
+
+  public ImmutableFSJobCatalog(Config sysConfig)
+      throws IOException {
+    this.sysConfig = sysConfig;
+    Properties sysProp = ConfigUtils.configToProperties(this.sysConfig);
+    Preconditions.checkArgument(sysProp.containsKey(ConfigurationKeys.JOB_CONFIG_FILE_GENERAL_PATH_KEY));
+
+    this.jobConfDirPath = new Path(sysProp.getProperty(ConfigurationKeys.JOB_CONFIG_FILE_GENERAL_PATH_KEY));
+    this.fs = this.jobConfDirPath.getFileSystem(new Configuration());
+    this.listeners = new JobCatalogListenersList(Optional.of(LOGGER));
+
+    this.loader = new PullFileLoader(jobConfDirPath, jobConfDirPath.getFileSystem(new Configuration()),
+        FSJobCatalogHelper.getJobConfigurationFileExtensions(sysProp),
+        PullFileLoader.DEFAULT_HOCON_PULL_FILE_EXTENSIONS);
+    this.converter = new FSJobCatalogHelper.JobSpecConverter(this.jobConfDirPath);
+  }
+
+  /**
+   * Fetch all the job files under the jobConfDirPath
+   * @return A collection of JobSpec
+   */
+  @Override
+  public synchronized List<JobSpec> getJobs() {
+    return Lists.transform(
+        Lists.newArrayList(loader.loadPullFilesRecursively(loader.getRootDirectory(), this.sysConfig, true)),
```

Comment:
The last argument (true) should be a call to a method e.g. shouldLoadGlobalConf(), so that the mutable subclass can override this to false.